### PR TITLE
[JioSaavn] Add extractor

### PR DESCRIPTION
**Pull request by awalgarg**
### Description of your *pull request* and other information

JioSaavn is a large streaming service primarily serving Indian music.

Original PR in youtube-dl: https://github.com/ytdl-org/youtube-dl/pull/20870, unfortunately it wasn't merged back then. I've been maintaining it in my personal fork since and it works well.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))

</details><details><summary>Copilot Summary</summary>  

### Summary
🎵🇮🇳🆕

Add support for JioSaavn streaming service. Add `jiosaavn.py` module with extractors for songs and albums. Import these extractors in `_extractors.py`.

> _`yt-dlp` expands_
> _JioSaavn songs and albums_
> _now in autumn breeze_

### Walkthrough
*  Add support for JioSaavn streaming service ([link](https://github.com/yt-dlp/yt-dlp/pull/8307/files?diff=unified&w=0#diff-780b22dc7eb280f5a7b2bbf79aff17826de88ddcbf2fc1116ba19901827aa4e3R896-R899), [link](https://github.com/yt-dlp/yt-dlp/pull/8307/files?diff=unified&w=0#diff-c1d315ae82c3ca04b6e2ddb15bace3e3e39dd1312f207df5dae69835238a28b5R1-R98))
  - Import JioSaavnSongIE and JioSaavnAlbumIE classes from `jiosaavn.py` module in `_extractors.py` ([link](https://github.com/yt-dlp/yt-dlp/pull/8307/files?diff=unified&w=0#diff-780b22dc7eb280f5a7b2bbf79aff17826de88ddcbf2fc1116ba19901827aa4e3R896-R899))
  - Define JioSaavnBaseInfoExtractor class to extract initial JSON data from webpage ([link](https://github.com/yt-dlp/yt-dlp/pull/8307/files?diff=unified&w=0#diff-c1d315ae82c3ca04b6e2ddb15bace3e3e39dd1312f207df5dae69835238a28b5R1-R98))
  - Define JioSaavnSongIE class to extract information and media URL of a single song ([link](https://github.com/yt-dlp/yt-dlp/pull/8307/files?diff=unified&w=0#diff-c1d315ae82c3ca04b6e2ddb15bace3e3e39dd1312f207df5dae69835238a28b5R1-R98))
  - Define JioSaavnAlbumIE class to extract information and URLs of all songs in an album ([link](https://github.com/yt-dlp/yt-dlp/pull/8307/files?diff=unified&w=0#diff-c1d315ae82c3ca04b6e2ddb15bace3e3e39dd1312f207df5dae69835238a28b5R1-R98))
  - Add tests and documentation for the new classes ([link](https://github.com/yt-dlp/yt-dlp/pull/8307/files?diff=unified&w=0#diff-c1d315ae82c3ca04b6e2ddb15bace3e3e39dd1312f207df5dae69835238a28b5R1-R98))

</details>

### CODE REVIEW

File: yt_dlp/extractor/jiosaavn.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            'title': song_data['title']['text'],
+            'formats': [{
+                'url': media_url,
+                'ext': 'mp3',
+                'format_note': 'MPEG audio',
+                'format_id': '128',
+                'vcodec': 'none',
+            }],
+            'album': song_data.get('album', {}).get('text'),
+            'thumbnail': song_data.get('image', [None])[0],
```
Comment (seproDev) [inline yt_dlp/extractor/jiosaavn.py:72]: I got returned a mp4 file with AAC audio, so hard-coding mp3 doesn't make sense. We can also make title extraction non-fatal.
I am a bit unsure regarding the `transform_source=clean_api_json`. Is that really needed? The two test cases seem to work without. I have removed it in the suggestion.
```suggestion
        media_data = self._download_json('https://www.jiosaavn.com/api.php', audio_id, data=data)

        return {
            'id': audio_id,
            'formats': [{
                'url': media_data['auth_url'],
                'container': media_data.get('type'),
                'vcodec': 'none',
            }],
            **traverse_obj(song_data, {
                'title': ('title', 'text'),
                'album': ('album', 'text'),
                'thumbnail': ('image', 0, {url_or_none}),
            }),
```
Tests will need to be adjusted and `url_or_none` and `from ..utils.traversal import traverse_obj` added to the imports.

File: yt_dlp/extractor/jiosaavn.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    }]
+
+    def _real_extract(self, url):
+        album_id = self._match_id(url)
+        json_data = self.extract_initial_data_as_json(url, album_id)
+        album_data = json_data['albumView']['album']
+        songs = json_data['albumView']['modules'][0]['data']
+        songs = [self.url_result(f'https://www.jiosaavn.com{song["title"]["action"]}') for song in songs]
+
+        return self.playlist_result(songs, album_id, album_data.get('title', {}).get('text'))
```
Comment (seproDev) [inline yt_dlp/extractor/jiosaavn.py:98]:
```suggestion
        title = traverse_obj(json_data, ('albumView', 'album', 'title', 'text'))
        songs = traverse_obj(
            json_data, ('albumView', 'modules', lambda _, x: x['key'] == 'list', 'data', ...,
                        'title', 'action', {lambda x: f'https://www.jiosaavn.com{x}'}, {self.url_result}))
        return self.playlist_result(songs, album_id, title)
```
`title` and `songs` could also be inlined.

**Review by seproDev — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/jiosaavn.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        'info_dict': {
+            'id': 'buIOjYZDrNA_',
+            'title': '96',
+        },
+        'playlist_count': 10,
+    }]
+
+    def _real_extract(self, url):
+        album_id = self._match_id(url)
+        json_data = self.extract_initial_data_as_json(url, album_id)
```
Comment (bashonly) [inline yt_dlp/extractor/jiosaavn.py:93]:
```suggestion
        album_id = self._match_id(url)
        album_view = self._extract_initial_data(url, album_id)['album_view']
```

File: yt_dlp/extractor/jiosaavn.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
+from .common import InfoExtractor
+from ..utils import (
+    js_to_json,
+    urlencode_postdata,
+)
```
Comment (bashonly) [inline yt_dlp/extractor/jiosaavn.py:5]:
```suggestion
    urlencode_postdata,
    urljoin,
)
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/jiosaavn.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+from ..utils import (
+    js_to_json,
+    urlencode_postdata,
+    urljoin,
+)
+
+
+class JioSaavnBaseInfoExtractor(InfoExtractor):
+    def _extract_initial_data(self, url, audio_id):
+        webpage = self._download_webpage(url, id)
```
Comment (bashonly) [inline yt_dlp/extractor/jiosaavn.py:11]:
```suggestion
        webpage = self._download_webpage(url, audio_id)
```

File: yt_dlp/extractor/jiosaavn.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
+from .common import InfoExtractor
+from ..utils import (
+    js_to_json,
+    urlencode_postdata,
+    urljoin,
+)
+
+
+class JioSaavnBaseInfoExtractor(InfoExtractor):
```
Comment (bashonly) [inline yt_dlp/extractor/jiosaavn.py:9]:
```suggestion
class JioSaavnBaseIE(InfoExtractor):
```

File: yt_dlp/extractor/jiosaavn.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    def _extract_initial_data(self, url, audio_id):
+        webpage = self._download_webpage(url, audio_id)
+        return self._search_json(
+            r'window\.__INITIAL_DATA__\s*=', webpage,
+            'init json', audio_id, transform_source=js_to_json)
+
+
+class JioSaavnSongIE(JioSaavnBaseInfoExtractor):
```
Comment (bashonly) [inline yt_dlp/extractor/jiosaavn.py:17]:
```suggestion
class JioSaavnSongIE(JioSaavnBaseIE):
```

File: yt_dlp/extractor/jiosaavn.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            }],
+            **traverse_obj(song_data, {
+                'title': ('title', 'text'),
+                'album': ('album', 'text'),
+                'thumbnail': ('image', 0, {url_or_none}),
+            }),
+        }
+
+
+class JioSaavnAlbumIE(JioSaavnBaseInfoExtractor):
```
Comment (bashonly) [inline yt_dlp/extractor/jiosaavn.py:60]:
```suggestion
class JioSaavnAlbumIE(JioSaavnBaseIE):
```

File: yt_dlp/extractor/jiosaavn.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
+from .common import InfoExtractor
+from ..utils import (
+    js_to_json,
+    urlencode_postdata,
+    urljoin,
+)
+
```
Comment (bashonly) [inline yt_dlp/extractor/jiosaavn.py:7]:
```suggestion
)
from ..utils.traversal import traverse_obj

```

File: yt_dlp/extractor/jiosaavn.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
+from .common import InfoExtractor
+from ..utils import (
+    js_to_json,
+    urlencode_postdata,
```
Comment (bashonly) [inline yt_dlp/extractor/jiosaavn.py:4]:
```suggestion
    js_to_json,
    url_or_none,
    urlencode_postdata,
```

File: yt_dlp/extractor/jiosaavn.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                'url': song_data['encrypted_media_url'],
+            }))
+
+        return {
+            'id': audio_id,
+            'formats': [{
+                'url': media_data['auth_url'],
+                'ext': media_data.get('type'),
+                'vcodec': 'none',
+            }],
```
Comment (bashonly) [inline yt_dlp/extractor/jiosaavn.py:53]:
```suggestion
            'url': media_data['auth_url'],
            'ext': media_data.get('type'),
            'vcodec': 'none',
```

**Comment by awalgarg** [comment]
Thanks for fixing it up and merging! I've been traveling so couldn't take a look earlier.

**Pull request merged**